_posts/2020-11-23-how-fast-are-ractors.md

```diff
@@ -0,0 +1,133 @@
+    <tbody>
+        <tr> <th>fork_test</th><td>16.44</td><td>0.175</td><td>16.84</td><td>0.181</td> </tr>
+        <tr> <th>thread_test</th><td>30.32</td><td>0.225</td><td>30.74</td><td>0.227</td> </tr>
+        <tr> <th>ractor_test</th><td>29.86</td><td>0.625</td><td>30.31</td><td>0.624</td> </tr>
+        <tr> <th>pipeless_ractor_test</th><td>24.48</td><td>0.278</td><td>24.98</td><td>0.277</td> </tr>
+    </tbody>
+</table>
+
+So: Ractors can be a bit ticklish. The plain (with-pipe) Ractor test never really got faster than threads. But the pipeless Ractor test ***does*** come out faster than threads and slower than fork, roughly as you'd expect.
+
+Still, even in this best case I measured: 16% faster than threads, but 33% slower than multiprocess. So: the dream of Ruby with full-speed native threads hasn't quite arrived yet.
+
+## Does More Work Per Chunk Help?
+
+I did a quick and dirty test with 10x as many numbers for each chunk of work for both threads and the pipeless Ractor test. After all, this is where Ractors should beat threads &mdash; lots of work that can be done on 4 CPUs for Ractors, but only one for threads. How did that do?
+
+Worse, actually. For whatever reason that version took 24.5 seconds for Ractors, but 22.1 seconds for threads (this is with 5 workers and 2,000 messages/worker.) I know Ractors have some odd interactions with the Ruby VM &mdash; Ruby isn't designed for separate per-thread garbage collection, for instance, as the JVM is. But running more calculation per message doesn't seem to be a clear win for Ractors as you'd expect it to be.
+
+Or maybe using exactly as many Ractors as CPUs? Nope, the quick-and-dirty test there came out at 9.3 seconds for Ractors vs 7.7 seconds for threads (2 workers, 2000 msgs/worker.)
+
+These additional tests show that the good result above was a bit of an anomaly. Not every setup benefits from Ractors yet, not even the ones you'd reasonably expect to. Again, we'll see how that changes as we near Ruby 3's release date.
+
+## So What's It All Mean?
+
+Ractors are quite new. Ruby 3 isn't due to be released (as I write this) for another month. I'm sure there will be both performance and stability work on Ractors before their final release.
+
+But initially, my recommendation would be that Ractors are finicky, but can give you performance more like native threads &mdash; like threads without the GVL. They're going to need some tuning, and I do ***not*** recommend them on Mac currently. But they look promising, and they seem to do what we've heard they will.
+
+My ***best test case gave Ractors a 16% advantage on two cores over regular Ruby threads***. So right now, the overhead is still high. Ruby still likes multiprocess (fork) concurrency, and Ractors aren't going to change that overnight.
+
+But Ractors do what they say they do, at least in prerelease Ruby. Keep in mind that they're only in Ruby 3. So you'll either need to use a prerelease Ruby, or wait until Christmas of 2020.
```

Comment:
today matz announced the event, maybe we can link that here?

https://rhc.connpass.com/event/195483

It's in japanese though haha.

Reply:
That's a fun idea! I've added it.